pkg/lock/sortable_mutex.go

```diff
@@ -0,0 +1,64 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package lock
+
+import (
+	"sort"
+	"sync"
+	"sync/atomic"
+)
+
+// sortableMutexSeq is a global sequence counter for the creation of new
+// SortableMutex's with unique sequence numbers.
+var sortableMutexSeq atomic.Uint64
+
+// SortableMutex provides a Mutex that can be globally sorted with other
+// sortable mutexes. This allows deadlock-safe locking of a set of mutexes
+// as it guarantees consistent lock ordering.
+type SortableMutex struct {
+	*sync.Mutex
+	seq uint64
+}
+
+// SortableMutexes is a set of mutexes that can be locked in a safe order.
+// Once Lock() is called it must not be mutated!
+type SortableMutexes []SortableMutex
+
+// Len implements sort.Interface.
+func (s SortableMutexes) Len() int {
+	return len(s)
+}
+
+// Less implements sort.Interface.
+func (s SortableMutexes) Less(i int, j int) bool {
+	return s[i].seq < s[j].seq
+}
+
+// Swap implements sort.Interface.
+func (s SortableMutexes) Swap(i int, j int) {
+	s[i], s[j] = s[j], s[i]
+}
+
+func (s SortableMutexes) Lock() {
```

Comment:
I think this deserves a comment to the effect of

Lock sorts the mutexes, and then locks them in order. If any lock cannot be acquired, this will block while holding the locks with a lower sequence number.

Reply:
Thanks, added!